File: src/122_Best_Time_to_Buy_and_Sell_Stock_II.rs
// 122. Best Time to Buy and Sell Stock II
// https://leetcode.cn/problems/best-time-to-buy-and-sell-stock-ii/

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut ans = 0;
        let mut new_prices = prices.clone();
        for i in 1..prices.len()-1 {
            if ((prices[i-1] < prices[i]) && (prices[i] < prices[i+1])) || ((prices[i-1] > prices[i]) && (prices[i] > prices[i+1])) {
                new_prices[i] = -1;
            }
        }
        let new_price: Vec<_> = new_prices.iter().filter(|&x| *x != -1).collect();
        // println!("{:?}", new_price);
        for i in 1..new_price.len() {
            let add = new_price[i] - new_price[i-1];
            if add >= 0 {
                ans += add;
            }
        }
        
        ans
    }
}

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
      let mut max_profit = 0;
        if prices.len() < 2 {
            return max_profit;
        }
        let mut prev_price = prices[0];
        for i in 1..prices.len() {
            if prices[i] > prev_price {
                max_profit += prices[i] - prev_price;
            }
            prev_price = prices[i];
        }
        max_profit

    }
}

File: src/122_Best_Time_to_Buy_and_Sell_Stock_II.py
# 122_Best_Time_to_Buy_and_Sell_Stock_II
# https://leetcode.cn/problems/best-time-to-buy-and-sell-stock-ii/

# 给你一个整数数组 prices ，其中 prices[i] 表示某支股票第 i 天的价格。

# 在每一天，你可以决定是否购买和/或出售股票。你在任何时候 最多 只能持有 一股 股票。你也可以先购买，然后在 同一天 出售。

# 返回 你能获得的 最大 利润 。


from typing import List
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ans = 0
        for i in range(len(prices)-1):
            if prices[i+1] - prices[i] >= 0:
                ans += (prices[i+1] - prices[i])
        
        return ans